doc/decisions/1_in_discussion/spec_expressiveness.md

```diff
@@ -2,22 +2,61 @@
 
 ## Problem
 
-Currently, you can easily come into wrong assumptions that something would work in the specification.
+The current implementation of the spec plugin is copying metadata to all other namespaces.
+Metadata e.g. `type` or `description` describes a certain key.
 
-We need to find minimal requirements to implement a sane spec plugin.
+Sometimes there are keys with similar or identical specifications.
+We want to define those in a single place and not have to manually copy metadata.
+`Globbing` expressions in key names comes into place at that time. 
+By providing certain wildcard characters e.g. `_` or `#` one can match multiple keys.
+This removes the need to explicitly define specification for keys.
+
+To describe the use case where `globbing` expressions make sense, we will use a REST application.
+This REST application uses multiple services which the user is not aware of at time of configuration.
+It could use a database, message broker, etc. but this is not yet known.
+A solution to this is `globbing` i.e. using wildcards in key names.
+The thing to be aware of is to define configuration keys which are necessary for every service.
+In this example we use `port` and `url`.
+
+The parent key for our specification is `/sw/org/app-name/#0/development`.
+
+```ni
+[service/_/port]
+meta:/require = true
+meta:/default = 8080
+meta:/description = "The port of the service."
+
+[service/_/url]
+meta:/require = true
+meta:/description = "The url of the service."
+```
+The `_` would match every `service` we are using in our application e.g. `database`, `message broker`, etc.
+
+But the current implementation has some undefined behaivour when using `globbing`.
+
+This specification uses `_` but we also have `#` as wildcard in `globbing`. 
```

Comment:
I would first explain the issues with `_` alone, i.e. that we cannot know the keynames that will be used with `ksLookup`, so we cannot create keys or copy metadata in advance.

Then you could mention that `#` also exists, and that because it is restricted to arrays it does _not_ have this problem. We only need to know the array size to create keys and copy metadata. We can simply enforce that the array size is defined.

Only then would I mentioned that `#` doesn't solve all problems. Overlapping specification is still and issue, but that's easy to fix. I'd just mention that any solution must report errors for overlapping specifications, something the current `spec` fails to do.

Finally, you should also mention the disadvantages of using `#` over `_`. Take your example, but with `#`:

```ni
[service/#/port]
meta:/require = true
meta:/default = 8080
meta:/description = "The port of the service."

[service/#/url]
meta:/require = true
meta:/description = "The url of the service."
```

Here we loose the unique name for the service that was previously covered by the `_` part. Now we just get an index in the array. Still a unique identifier, but much harder to handle. Also it's not easily possible to give meaning to this index. Whereas before using `database` for `_` could indicate something to the application (and the user), you now only have something like `service/#0/port`. You can't really give meaning to `#0`.

That's why you'll often end up with things like:

```ni
[service/#/port]
meta:/require = true
meta:/default = 8080
meta:/description = "The port of the service."

[service/#/url]
meta:/require = true
meta:/description = "The url of the service."

[service/#/name]
meta:/default = ""
meta:/description = "The name of the service."
```

Now we have a name again. But that is still more annoying to use. You have to set more keys. The key `service/#0/url` still doesn't tell you much. You still have to look at `service/#0/name` to find out what the URL is for, etc.

That is why we would like to have a `_` that works at least somewhat decently.

> **Note**: Using `service/#/name` to define what kind of service this array entry is, also has an advantage over using the `_` part for this. The value of `service/#/name` can be restricted with a `meta:/check/enum`, while `_` is completely arbitrary. The tags idea in #3954 could address that.